### PR TITLE
Polydisperse neighbor search algorithm

### PR DESCRIPTION
**Pull request by AMLattanzi**
## Summary
New functionality for building the particle neighbor list with multiple grids. The multi-grid approach can provide significant speed up with polydisperse particle-laden flows.

## Additional background
Algorithm based upon: 
Shire, T., Hanley, K.J. & Stratford, K. DEM simulations of polydisperse media: efficient contact detection applied to investigate the quasi-static limit. Comp. Part. Mech. 8, 653–663 (2021). https://doi.org/10.1007/s40571-020-00361-2

## Checklist
1. Verify correctness with standard single-grid (SG) neighbor search (MFiX regtests).
2. Test speed with SG neighbor search (Perlmutter w/ canonical sim)
3. Verify new multi-grid (MG) gives same results as SG (new polydisperse test)
4. Document speed up of MG vs SG for polydisperse test

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

File: Src/Particle/AMReX_DenseBins.H [hunk 1]
```diff
@@ -50,6 +50,26 @@ struct DenseBinIteratorFactory
     const T* m_items;
 };
 
+// AML
```
Comment (atmyers) [inline Src/Particle/AMReX_DenseBins.H]: Could you remove the `// AML` comments for the PR? Git will show you as the most recent to edit these lines, so there's no need to also label them.

File: Src/Particle/AMReX_NeighborList.H [hunk 2]
```diff
@@ -185,140 +202,186 @@ class NeighborList
...
+        auto plo_p      = plo_v.data();
+        auto  lo_p      =  lo_v.data();
+        auto  hi_p      =  hi_v.data();
+        BinMapper bm(off_bins_p, dxi_p, plo_p, lo_p, hi_p, bin_type_array);
 
-        // first pass - count the number of neighbors for each particle
-        const size_t np_real  = src_tile.numRealParticles();
-        m_nbor_counts.resize(np_real+1, 0);
-        m_nbor_offsets.resize(np_real+1);
+        // TODO: bin_type_array will live inside BinMapper (f)
```
Comment (atmyers) [inline Src/Particle/AMReX_NeighborList.H]: Could you remove this comment as well?

**Comment by AMLattanzi** [comment]
Absolutely; that was just for me to grep where I had touched stuff.

**Comment by AMLattanzi** [comment]
Yes.

File: Src/Particle/AMReX_Particle.H [hunk 3]
```diff
@@ -424,16 +424,19 @@ std::ostream&
 operator<< (std::ostream& os, const Particle<NReal, NInt>& p)
 {
     os << p.id()   << ' '
-       << p.cpu()  << ' ';
+       << p.cpu()  << "\n";
```
Comment (atmyers) [inline Src/Particle/AMReX_Particle.H]: Could you revert the changes to this file? I think this might cause some of the regression tests for some codes that look at ASCII to fail.

**Comment by atmyers** [comment]
Looks good, I left a few minor comments. Could you try running an MFIX problem on Perlmutter or somewhere to verify that these changes don't slow down the neighbor list construction time in the non-polydisperse case?

**Review by atmyers — APPROVED** [review]

**Comment by atmyers** [comment]
This looks good to me, thank you for adding this capability!
